app/services/katello/event_daemon/monitor.rb

```diff
@@ -0,0 +1,54 @@
+module Katello
+  module EventDaemon
+    class Monitor
+      def initialize(service_classes)
+        @service_classes = service_classes
+      end
+
+      def start
+        error = nil
+        status = nil
+        loop do
+          Rails.application.executor.wrap do
+            check_services(error, status)
+          end
+          sleep 15
+        end
+      end
+
+      def check_services(error, status)
+        @service_classes.each do |service_class|
+          status = service_class.status
+        rescue => error
+          Rails.logger.error("Error occurred while pinging #{service_class}")
+          Rails.logger.error(error.message)
+          Rails.logger.error(error.backtrace.join("\n"))
+        ensure
+          if error || !status&.dig(:running)
+            begin
+              Rails.logger.info "******************* RESTARTING #{service_class}"
+              service_class.close
+              service_class.run
+              service_class.status
+            rescue => error
+              Rails.logger.error("Error occurred while starting #{service_class}")
+              Rails.logger.error(error.message)
+              Rails.logger.error(error.backtrace.join("\n"))
+            ensure
+              error = nil
+            end
+          end
+          write_statuses_to_cache
```

Comment:
This implementation means that we are calling .status on all the service classes an additional time. What about building the entire status during the execution of `check_statuses`?

for example:

```
def initialize(...)
   ...
  @service_statuses = {}
end
```

near the top of check_services:
```
  statuses[service_name] = service_class.status
````

and at the end

```
Rails.cache.write(KEY, @service_statuses)
```